// sqreen/test/lib/instrumentation/instrumentationInterface.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const InstrumentationInterface = require('../../../lib/instrumentation/instrumentationInterface');
const NS = require('../../../lib/instrumentation/hooks/ns');

describe('InstrumentationInterface', () => {

    it('should cover wrap', { plan: 4 }, (done) => {

        const mod = {
            a: function (x) {

                return x + 1;
            }
        };
        const method = function (args) {

            expect(args[0]).to.equal(1);
        };
        method.noBudget = true;
        const patch = InstrumentationInterface.strategies.wrap(mod, 'a', {}, 'mmmod', { method });
        const patch2 = InstrumentationInterface.strategies.wrap(mod, 'a', {}, 'mmmod', { method });
        expect(patch).to.equal(patch2);
        expect(mod.a(1)).to.equal(2);
        done();
    });

    it('should cover a masswrap', { plan: 5 }, (done) => {

        const exp = {
            val: 'foo'
        };
        const module = {
            action: function (str) {

                expect(str).to.equal(exp.val);
            }
        };
        let call = 0;
        InstrumentationInterface.strategies.massWrap(module, ['action'], (self, args) => {

            ++call;
            if (call === 1) {
                exp.val = 'foo';
                throw new Error();
            }
            if (call === 2) {
                exp.val = 'bar';
                args[0] = 'bar';
            }
            if (call === 3) {
                exp.val = 'foo';
            }
        });
        module.action('foo');
        module.action('foo');
        const ns = NS.getNS();
        ns.run(() => {

            ns.set('budget', {
                stopCount: function () {

                    expect(1).to.equal(1);
                },
                startCount: function () {

                    expect(1).to.equal(1);
                }
            });
            module.action('foo');
            ns.set('budget', undefined);
            done();
        });
    });
});
