// sqreen/test/lib/instrumentation/hooks/knex.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;
const beforeEach = lab.beforeEach;

const Hook = require('../../../../lib/instrumentation/hooks/knexHook');
const NS = require('../../../../lib/instrumentation/hooks/ns').getNS();

describe('knex hook', () => {

    beforeEach((done) => {

        try {
            NS.exit(NS.active);
        }
        catch (_) {}
        done();
    });

    it('should test the hook', { plan: 2 }, (done) => {

        let i = 0;
        const Client = class {

            queryBuilder() {

                return {
                    then: function (cb) {

                        return cb();
                    }
                };
            }

            runner() {

                return {
                    client: {
                        releaseConnection: function () {

                            i++;
                            if (i === 2) {
                                done();
                            }
                        }
                    }
                };
            }
        };

        const mod = {
            Client
        };

        Hook({}, mod);

        const client = new Client();
        client.runner();
        NS.run(() => {

            NS.set('a', 'hello');

            client.queryBuilder()
                .then(() => {

                    expect(NS.get('a')).to.equal('hello');
                });

            client.runner();
            expect(NS.get('raise').size).to.equal(1);

            NS.get('raise').get('knex.onraise')();
            client.runner();
            NS.get('raise').get('knex.onraise')();
        });
    });

    it('should test the hook and release connection', { plan: 3 }, (done) => {

        const Client = class {

            queryBuilder() {

                return {
                    then: function (cb) {

                        return cb();
                    }
                };
            }

            releaseConnection() {

            }

            runner() {

                return {
                    client: this
                };
            }
        };

        const mod = {
            Client
        };

        Hook({}, mod);

        const client = new Client();
        client.runner();
        client.releaseConnection();
        NS.run(() => {

            NS.set('a', 'hello');

            client.queryBuilder()
                .then(() => {

                    expect(NS.get('a')).to.equal('hello');
                });
            client.releaseConnection();
            client.runner();
            expect(NS.get('raise').size).to.equal(1);
            client.releaseConnection();
            expect(NS.get('raise').size).to.equal(0);
            done();
        });
    });
});
