// sqreen/test/lib/metric/average.js
/**
 * Copyright (c) 2016 - 2020 Sqreen. All Rights Reserved.
 * Please refer to our terms for more information: https://www.sqreen.io/terms.html
 */
'use strict';
const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();
const beforeEach = lab.beforeEach;

const describe = lab.describe;
const it = lab.it;
const expect = Code.expect;

const Metric = require('../../../lib/metric');
const Avg = require('../../../lib/metric/average');

const SqreenSDK = require('sqreen-sdk');

describe('Metric', () => {

    beforeEach((done) => {

        Metric._clearAllMetrics();
        require('../../../lib/command/features').featureHolder.use_signals = true;
        done();
    });

    describe('Avg', () => {

        it('should create an average metric', { plan: 4, timeout: 20000 }, (done) => {

            Metric.getAllReports(true);
            SqreenSDK.initBatch(10000, 1000, () => Promise.resolve());
            const avg = new Avg({
                kind: 'Average',
                name: 'login-fail',
                period: 2
            });

            avg.add('key1', 0);
            avg.add('key1', 2);
            avg.add('key2', 1);
            avg.add(null, 1);
            avg.add({ a: 1 }, 1);
            avg.add({ a: 1 }, 1);
            avg.add({ a: 1 }, 1);

            expect(Metric.getMetricByName('login-fail')).to.equal(avg);
            setTimeout(() => {

                avg.add('key1', 10);
                avg.add('key1', 5);
                avg.add('key3', 1);
            }, 3000);

            setTimeout(() => {

                Metric.getAllReports(true);
                const metrics = SqreenSDK.Signal.prototype.BATCH.data;
                expect(metrics).to.have.length(2);
                expect(metrics[0].payload.values).to.equal([{ key: 'key1', value: 1 },{ key: 'key2', value: 1 }, { key: { a: 1 }, value: 1 }]);
                expect(metrics[1].payload.values).to.equal([{ key: 'key1', value: 7.5 },{ key: 'key3', value: 1 }]);
                done();
            }, 6000);

        });
    });

});
